### ISSUE TITLE
Why doesn't the callback for <For /> provide a reference to the original collection?

### ISSUE THREAD

**Issue by promethyttrium** [solidjs/solid#1779]
Ideally, I should be able to do this:
```
<For each={Object.keys(someVeryComplexOperationThatReturnsAMap())>
  {(key, index, results ) => <SomeDisplayComponent item={results[key]} />} 
</For>
```
But I can't because there isn't a 3rd parameter that references the result...

**Comment by atk** [comment]
This is one of the cases where performance is at odds with DX: adding the ref means that the reference is kept in the closure; any asynchronous process running within will keep the closure alive, needlessly increasing memory consumption.

In your example case, consider using createMemo for the very complex operation, so that you can directly reference its output from wherever you want without adding to the memory.

**Comment by ryansolid** [comment]
Yeah never thought about adding it. It's interesting because to be reactive I can't just wrap the value but wrap the access so in a sense I'd just be passing the same closure through so it isn't that expensive on the basic level. But it could turn something that didn't require to be memoized to be so because you would suddenly be redoing that `Object.keys` on every single row. So the solution would be that we auto-memoize the expression. At which point by not providing the API we are directing you to do that yourself without opting everyone into automemoization.

I think that this is an interesting suggestion and is probably more of a disccusion.
